### ISSUE TITLE
[API Proposal]: Mark the return value of `ILogger.BeginScope` MaybeNull

### ISSUE THREAD

**Issue by skyoxZ** [dotnet/runtime#63867]
### Background and motivation

The return value of [ILogger.BeginScope](https://github.com/dotnet/runtime/blob/v6.0.1/src/libraries/Microsoft.Extensions.Logging.Abstractions/ref/Microsoft.Extensions.Logging.Abstractions.cs#L31) is marked `NotNull` since #43892.

However, I have seen many tutorials (e.g. [Microsoft doc](https://docs.microsoft.com/en-us/dotnet/core/extensions/custom-logging-provider)) just let it return null so I suggest marking the return value `MaybeNull`.

And another option is making [NullScope](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/src/Extensions/Logging/NullScope.cs) public which is quite useful when implementing a custom logger.

 

### API Proposal

```C#
namespace Microsoft.Extensions.Logging
{
    public interface ILogger
    {
        ...
        System.IDisposable? BeginScope<TState>(TState state);
    }
}
```

### API Usage

-

**Comment by Joe4evr** [comment]
Does a null `IDisposable` work properly in a `using` block? 💭

**Comment by alexrp** [comment]
> Does a null `IDisposable` work properly in a `using` block? 💭

Yes, a null check is inserted in the generated `finally` block.

**Comment by skyoxZ** [comment]
[Api doc](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.loggerextensions.beginscope?view=dotnet-plat-ext-6.0) says the return value "can be null". Maybe it's just because the doc is outdating but it's true that returning null will not bring any error.

**Comment by FiniteReality** [comment]
I imagine the return value of ILogger.BeginScope is marked NotNull because the average user won't ever get a null scope from it. In fact, there's only ever one case where it could return null, and that's if only one logger has been specified, and it uses a custom implementation: <https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging/src/Logger.cs#L123>

**Comment by stephentoub** [comment]
If it's considered valid to return null even if it's rarely done, the return type should be nullable.

**Comment by Joe4evr** [comment]
> If it's considered valid to return null even if it's rarely done, the return type should be nullable.

😅 For a moment I was thinking about the deref warnings, but then I realized the return type is only `IDisposable` (and not an extended interface), so there's not really all that much for consumers to dot off for (without casting in the first place). Most user code will already look like:
```cs
using (logger.BeginScope()) { }
```
rather than:
```cs
using (var scope = logger.BeginScope()) { }
```
so I believe it'd be very rare for code to actually *receive* the warnings despite being tracked.

**Comment by eerhardt** [comment]
Fixed by #66960.
